java/ql/lib/semmle/code/java/frameworks/android/Notifications.qll

```diff
@@ -0,0 +1,55 @@
+/** Provides classes and predicates related to Android notifications. */
+
+import java
+private import semmle.code.java.dataflow.DataFlow
+private import semmle.code.java.dataflow.ExternalFlow
+private import semmle.code.java.dataflow.FlowSteps
+
+private class NotificationBuildersSummaryModels extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        "android.app;Notification$Action;true;Action;(int,CharSequence,PendingIntent);;Argument[2];Argument[-1];taint",
+        "android.app;Notification$Action$Builder;true;Builder;(int,CharSequence,PendingIntent);;Argument[2];Argument[-1];taint",
+        "android.app;Notification$Action$Builder;true;Builder;(Icon,CharSequence,PendingIntent);;Argument[2];Argument[-1];taint",
+        "android.app;Notification$Action$Builder;true;Builder;(Action);;Argument[0];Argument[-1];taint",
+        "android.app;Notification$Action$Builder;true;addExtras;;;MapKey of Argument[0];MapKey of SyntheticField[android.content.Intent.extras] of Argument[-1];value",
+        "android.app;Notification$Action$Builder;true;addExtras;;;MapValue of Argument[0];MapValue of SyntheticField[android.content.Intent.extras] of Argument[-1];value",
+        "android.app;Notification$Action$Builder;true;build;;;Argument[-1];ReturnValue;taint",
+        "android.app;Notification$Action$Builder;true;getExtras;;;SyntheticField[android.content.Intent.extras] of Argument[-1];ReturnValue;value",
+        "android.app;Notification$Builder;true;addAction;(int,CharSequence,PendingIntent);;Argument[2];Argument[-1];taint",
+        "android.app;Notification$Builder;true;addAction;(Action);;Argument[0];Argument[-1];taint",
+        "android.app;Notification$Builder;true;addExtras;;;MapKey of Argument[0];MapKey of SyntheticField[android.content.Intent.extras] of Argument[-1];value",
+        "android.app;Notification$Builder;true;addExtras;;;MapValue of Argument[0];MapValue of SyntheticField[android.content.Intent.extras] of Argument[-1];value",
+        "android.app;Notification$Builder;true;build;;;Argument[-1];ReturnValue;taint",
```

Comment:
Are the two `build` methods supposed to keep the `extras` map? I.e. something like value step from `SyntheticField[android.content.Intent.extras] of Argument[-1]` to `SyntheticField[android.content.Intent.extras] of ReturnValue`? Or is the `extras` map no longer available on the results of `Builder.build`?

Reply:
You are right. `Notification$Builder.build` creates a `Notification`, which has an [`extras`](https://developer.android.com/reference/android/app/Notification#extras) field. `Notification$Action$Builder.build` creates a `Notification$Action`, which has the [`getExtras`](https://developer.android.com/reference/android/app/Notification.Action#getExtras()) method.

Added in 522a4bb.

Apparently, the `Action.extras` aren't merged with the `Notification.extras` of the notification the action is added to through `addAction`, so we shouldn't need an additional value step from `SyntheticField[android.content.Intent.extras] of Argument[0]` (the `Action`) to `Field[android.app.Notification.extras] of Argument[-1]` (the `Notification`).